services/supervisor/tests/supervisor/migrations/mod.rs

```diff
@@ -0,0 +1,879 @@
+}
+
+/// Same as `testkit_with_supervisor_and_service`, but services do not support migrations.
+fn testkit_with_supervisor_and_service_no_migrations(validator_count: u16) -> TestKit {
+    // Initialize builder;
+    let builder = TestKitBuilder::validator().with_validators(validator_count);
+
+    // Add supervisor.
+    let builder = builder
+        .with_rust_service(Supervisor)
+        .with_artifact(Supervisor.artifact_id())
+        .with_instance(Supervisor::simple());
+
+    // Add MigrationService with running instance.
+    let builder = builder.with_default_rust_service(MigrationService);
+
+    // Add migrating artifact for version 0.2.
+    let builder = builder
+        .with_rust_service(MigrationServiceV02)
+        .with_artifact(MigrationServiceV02.artifact_id());
+
+    // Add artifact for version 0.5.
+    let builder = builder
+        .with_rust_service(MigrationServiceV05)
+        .with_artifact(MigrationServiceV05.artifact_id());
+
+    builder.create()
+}
+
+/// Sends a `MigrationRequest` to supervisor through API.
+fn request_migration(api: &TestKitApi, request: MigrationRequest) -> Hash {
+    let hash: Hash = api
+        .private(ApiKind::Service("supervisor"))
+        .query(&request)
+        .post("migrate")
+        .unwrap();
+    hash
+}
+
+/// Obtains a migration state through API.
+fn migration_state(api: &TestKitApi, request: MigrationRequest) -> MigrationState {
+    let query: MigrationInfoQuery = request.into();
+    api.private(ApiKind::Service("supervisor"))
+        .query(&query)
+        .get("migration-status")
+        .unwrap()
+}
+
+/// Stops service with the given ID.
+fn stop_service(testkit: &mut TestKit, id: InstanceId) {
+    let keypair = testkit.us().service_keypair();
+    execute_transaction(
+        testkit,
+        ConfigPropose::immediate(0)
+            .stop_service(id)
+            .sign_for_supervisor(keypair.0, &keypair.1),
+    )
+    .expect("Stop service transaction should be processed");
+}
+
+fn obtain_expected_hash(testkit: &mut TestKit, request: &MigrationRequest) -> Hash {
+    for _ in 0..5 {
+        let snapshot = testkit.snapshot();
+        let prefixed = Prefixed::new(Supervisor::NAME, &snapshot);
+        let schema = SchemaImpl::new(prefixed);
+        let state = schema.migration_state_unchecked(&request);
+
+        assert!(
+            state.is_pending(),
+            "State changed from pending while awaiting for expected hash: {:?}",
+            state
+        );
+
+        let expected_hash = state.expected_state_hash();
+
+        if let Some(expected_hash) = expected_hash {
+            return *expected_hash;
+        } else {
+            // Migration is executed in the separate thread, so sleep a bit.
+            thread::sleep(Duration::from_millis(50));
+            // Then create a new block.
+            testkit.create_block();
+        }
+    }
+    panic!("Node didn't calculate the expected hash")
+}
+
+fn wait_for_migration_success(
```

Comment:
Nit: It could make sense to extract common code for `wait_for_migration_success`, `wait_for_migration_fail` and `wait_for_migration_timeout`, which would wait until the migration request status differs from `Pending`.